core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlockEncoding.java

```diff
@@ -40,14 +40,22 @@ public void writeBlock(BlockEncodingSerde blockEncodingSerde, SliceOutput sliceO
         AbstractVariableWidthBlock variableWidthBlock = (AbstractVariableWidthBlock) block;
 
         int positionCount = variableWidthBlock.getPositionCount();
+        int notNullPositionsCount = 0;
+        for (int i = 0; i < positionCount; i++) {
+            notNullPositionsCount += variableWidthBlock.isNull(i) ? 0 : 1;
+        }
+
         sliceOutput.appendInt(positionCount);
+        sliceOutput.appendInt(notNullPositionsCount);
 
         // offsets
         int totalLength = 0;
         for (int position = 0; position < positionCount; position++) {
             int length = variableWidthBlock.getSliceLength(position);
             totalLength += length;
-            sliceOutput.appendInt(totalLength);
+            if (!variableWidthBlock.isNull(position)) {
```

Comment:
Do you think it makes sense to fall back to the old method if there are zero nulls? this could have lower CPU usage.

Reply:
I also wonder if this fallback could be used is number of nulls is small e.g. 1%. Do you know what is the impact on CPU of this change?

Reply:
@lukasz-stec, it could be done as you suggest, but it should be decided with microbenchmark results. 
In macrobenchmark I do not observe regression of CPU time generally. I do not see regression for `tpch/q10`. This one query operates on varchars a bit.